src/webview/html/editHistoryHtml.js

```diff
@@ -0,0 +1,105 @@
+/* @flow strict-local */
+import template from './template';
+import type { ThemeName, MessageSnapshot, UserOrBot, Auth } from '../../types';
+import { shortTime, humanDate } from '../../utils/date';
+import css from '../css/css';
+import script from '../js/script';
+
+const TOPIC_EDITED = 'TOPIC_EDITED';
+const CONTENT_EDITED = 'CONTENT_EDITED';
+const BOTH_EDITED = 'BOTH_EDITED';
+const NOT_EDITED = 'NOT_EDITED';
+
+const renderSnapshot = (snapshot: MessageSnapshot, usersById: Map<number, UserOrBot>): string => {
+  const user = usersById.get(snapshot.user_id);
+  const date = new Date(snapshot.timestamp * 1000);
+  const editedTime = `${humanDate(date)} ${shortTime(date)}`;
+  let displayedContent: string = '';
+  let tag: string = '';
+  let edit_flag;
+
+  if (snapshot.prev_topic === undefined && snapshot.prev_content === undefined) {
+    edit_flag = NOT_EDITED;
+  } else if (snapshot.prev_content !== undefined && snapshot.prev_topic === undefined) {
+    edit_flag = CONTENT_EDITED;
+  } else if (snapshot.prev_content === undefined && snapshot.prev_topic !== undefined) {
+    edit_flag = TOPIC_EDITED;
+  } else {
+    edit_flag = BOTH_EDITED;
+  }
+
+  switch (edit_flag) {
+    case NOT_EDITED: {
+      tag += template`<span class="message-tag">Original message</span>`;
+      displayedContent = snapshot.content;
+      break;
+    }
+
+    case TOPIC_EDITED: {
+      tag += template`<span class="message-tag">Topic edited by ${
+        user ? user.full_name : 'unknown user'
+      }</span>`;
+
+      // prev_topic is always present for this case. $FlowFixMe.
+      displayedContent = template`Topic:
+      <span class="highlight_text_inserted">${snapshot.topic}</span>
+      <span class="highlight_text_deleted">${snapshot.prev_topic}</span>`;
+      break;
+    }
+
+    case CONTENT_EDITED: {
+      tag += template`<span class="message-tag">Edited by ${
+        user ? user.full_name : 'unknown user'
+      }</span>`;
+
+      // If prev_content is present, then so is content_html_diff. $FlowFixMe.
+      displayedContent = snapshot.content_html_diff;
+      break;
+    }
+
+    case BOTH_EDITED: {
+      tag += template`<span class="message-tag">Edited by ${
+        user ? user.full_name : 'unknown user'
+      }</span>`;
+
+      // prev_topic and content_html_diff are always present for this case. $FlowFixMe.
+      displayedContent = template`Topic:
+      <span class="highlight_text_inserted">${snapshot.topic}</span>
+      <span class="highlight_text_deleted">${snapshot.prev_topic}</span>
+      <br/>$!${snapshot.content_html_diff}`;
+      break;
+    }
+
+    default:
+      // eslint-disable-next-line no-console
+      console.error('Invalid edit_flag');
```

Comment:
Let's get this logged to Sentry, with `logging.error`, and include the value of `edit_flag` in the second argument, `extras`.

Reply:
Ok
(It's actually impossible to reach here :p)

Reply:
> (It's actually impossible to reach here :p)

Quick tip: you can add `ensureUnreachable(edit_flag);` here to assert this to the type-checker (and ask it to verify that it can prove it.) That helps prevent errors before they even reach users :slightly_smiling_face: if say a new case is added here in the future.

Reply:
I'm getting the following error on adding that:
```js
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/webview/html/editHistoryHtml.js:76:25

Cannot call ensureUnreachable with edit_flag bound to x because:
 • string [1] is incompatible with empty [2].
 • string [3] is incompatible with empty [2].
 • string [4] is incompatible with empty [2].
 • string [5] is incompatible with empty [2].

     src/webview/html/editHistoryHtml.js
 [1] 10│ const TOPIC_EDITED = 'TOPIC_EDITED';
 [3] 11│ const CONTENT_EDITED = 'CONTENT_EDITED';
 [4] 12│ const BOTH_EDITED = 'BOTH_EDITED';
 [5] 13│ const NOT_EDITED = 'NOT_EDITED';
       :
     73│     }
     74│
     75│     default:
     76│       ensureUnreachable(edit_flag);
     77│       logging.error('Invalid edit_flag', { editFlag: edit_flag });
     78│       break;
     79│   }

     src/types.js
 [2] 42│ export function ensureUnreachable(x: empty) {}



Found 4 errors
```

I think Flow is not able to understand that this case is unreachable?

Reply:
Ah -- it's that the type of `edit_flag` is simply `string`, instead of being an enum `'TOPIC_EDITED' | 'CONTENT_EDITED' | 'BOTH_EDITED' | 'NOT_EDITED'`.

You could make it an enum by annotating it with that type (and then probably doing something to avoid initializing to `undefined` and mutating.)

Or, looking a bit more at the code: this flag is so short-lived, just from an if/else-if chain to a `switch` right after it. I think it'd simplify things to just fuse those together: have an if/else-if chain with those conditions, and with the cases' bodies as their bodies. Or, a bit further: a two-level tree of if/else statements:
```js
if (snapshot.prev_content === undefined) {
  if (snapshot.prev_topic === undefined) {
  …
  } else {
  …
  }
} else {
  …
}